ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -996,12 +1014,76 @@ let rank_hosts_by_best_vgpu ~__context vgpu visible_hosts =
                0L
         )
         hosts
-      |> List.map (fun g -> List.map (fun (h, _) -> h) g)
+      |> List.map (fun g -> List.map fst g)
+
+(* Group all hosts to 2 parts:
+   1. A list of affinity host (only one host).
+   2. A list of lists, each list contains hosts with the same number of
+      running VM in that anti-affinity group.
+      These lists are sorted by VM's count.
+   Combine these lists into one list. The list is like below:
+   [ [host1] (affinity host)
+   , [host2, host3] (no VM running)
+   , [host4, host5] (one VM running)
+   , [host6, host7] (more VMs running)
+   , ...
+   ]
+*)
+let rank_hosts_by_placement ~__context ~vm ~group =
+  let affinity_host =
+    match Db.VM.get_affinity ~__context ~self:vm with
+    | ref when Db.is_valid_ref __context ref ->
+        [ref]
+    | _ ->
+        []
+  in
+  let sorted_hosts =
+    let host_without_affinity_host =
+      Db.Host.get_all ~__context
+      |> List.filter (fun host -> not (List.mem host affinity_host))
+    in
+    let host_tbl =
```

Comment:
I think the current code would work. However, a more functional version of this would use the `Map` module rather than a hash table. You can define one for refs:
```
module RefMap = Map.Make(Ref)
```
It has a handy `update` function to modify keys. Also you don't really need to initialise it with zeros.

Something like:
```
let host_of_vm vm =
  let vm_rec = Db.VM.get_record ~__context ~self:vm in
   match
   ...
in
let host_map =
  Db.VM_group.get_VMs ~__context ~self:group
  |> List.fold_left (fun m vm ->
      match host_of_vm vm with
      | Some h ->
          RefMap.update h
            (fun c -> Option.(value ~default:0 c |> succ |> some))
            m
      | None -> m
    ) RefMap.empty
in
... Helpers.group_by using RefMap.find_opt ...
```